worlds/zork_grand_inquisitor/data_funcs.py

```diff
@@ -0,0 +1,247 @@
+    items: Set[ZorkGrandInquisitorItems] = set()
+
+    item: ZorkGrandInquisitorItems
+    data: ZorkGrandInquisitorItemData
+    for item, data in item_data.items():
+        if data.tags is not None and tag in data.tags:
+            items.add(item)
+
+    return items
+
+
+def game_id_to_items() -> Dict[int, ZorkGrandInquisitorItems]:
+    mapping: Dict[int, ZorkGrandInquisitorItems] = dict()
+
+    item: ZorkGrandInquisitorItems
+    data: ZorkGrandInquisitorItemData
+    for item, data in item_data.items():
+        if data.statemap_keys is not None:
+            for key in data.statemap_keys:
+                mapping[key] = item
+
+    return mapping
+
+
+def location_groups() -> Dict[str, Set[str]]:
+    groups: Dict[str, Set[str]] = dict()
+
+    tag: ZorkGrandInquisitorTags
+    for tag in ZorkGrandInquisitorTags:
+        groups[tag.value] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if data.tags is not None:
+            for tag in data.tags:
+                groups[tag.value].add(location.value)
+
+    return {k: v for k, v in groups.items() if len(v)}
+
+
+def locations_by_region(include_deathsanity: bool = False) -> Dict[
+    ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]
+]:
+    mapping: Dict[ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]] = dict()
+
+    region: ZorkGrandInquisitorRegions
+    for region in ZorkGrandInquisitorRegions:
+        mapping[region] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if include_deathsanity is False and ZorkGrandInquisitorTags.DEATHSANITY in (
+            data.tags or tuple()
+        ):
+            continue
+
+        mapping[data.region].add(location)
+
+    return mapping
+
+
+def locations_with_tag(tag: ZorkGrandInquisitorTags) -> Set[ZorkGrandInquisitorLocations]:
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+
+    return {location for location, data in location_data.items() if data.tags is not None and tag in data.tags}
+
+
+def location_access_rule_for(location: ZorkGrandInquisitorLocations, player: int) -> str:
+    data: ZorkGrandInquisitorLocationData = location_data[location]
+
+    if data.requirements is None:
+        return "lambda state: True"
+
+    lambda_string: str = "lambda state: "
+
+    i: int
+    requirement: Union[
+        Tuple[
+            Union[
+                ZorkGrandInquisitorEvents,
+                ZorkGrandInquisitorItems,
+            ],
+            ...,
+        ],
+        ZorkGrandInquisitorEvents,
+        ZorkGrandInquisitorItems
+    ]
+
+    for i, requirement in enumerate(data.requirements):
+        if isinstance(requirement, tuple):
+            lambda_string += "("
+
+            ii: int
```

Comment:
They are the indices coming from using `enumerate`. It's extremely common in Python to do it like this. You can find multiple examples of core doing it exactly like this.

![image](https://github.com/ArchipelagoMW/Archipelago/assets/35039/78494b69-f2f5-441f-b481-a13a9280503d)

The `ii` is less common but it's how I've personally always handled 2D lists. Others prefer `i` `j` `k`. I prefer `ii` because it implies the level of nesting you are at and I can always tell it's meant to be an index at a glance